contracts/src/Gateway.sol

```diff
@@ -577,6 +568,29 @@ contract Gateway is IGateway, IInitializable {
         UD60x18 multiplier;
     }
 
+    function _upgrade(address impl, bytes32 implCodeHash, bytes memory initializerParams) internal {
+        // Verify that the implementation is actually a contract
+        if (!impl.isContract()) {
+            revert InvalidCodeHash();
+        }
+
+        // As a sanity check, ensure that the codehash of implementation contract
+        // matches the codehash in the upgrade proposal
+        if (impl.codehash != implCodeHash) {
+            revert InvalidCodeHash();
+        }
+
+        // Update the proxy with the address of the new implementation
+        ERC1967.store(impl);
+
+        // Call the initializer
+        (bool success, bytes memory returndata) =
+            impl.delegatecall(abi.encodeCall(IInitializable.initialize, initializerParams));
```

Comment:
If initializer is always run. Than we cannot through `AlreadyInitialized()` on line https://github.com/Snowfork/snowbridge/pull/1170/files#diff-74470ba8f35b42dbc8805739e92708c0696343e0b1819feb3e7a5ebb77cc623cR605

Wouldn't it need to skip if already initialized?

Reply:
Well the idea is that each new version of the Gateway will have a new initializer defined, like in `contracts/src/upgrades/rococo/GatewayV2.sol`.

Can you foresee any problems with that?

Reply:
Ah ok. I get the pattern now!